File: mobile/src/main/java/com/alexstyl/specialdates/service/CustomEventProvider.java
package com.alexstyl.specialdates.service;

import android.content.ContentResolver;
import android.database.Cursor;
import android.net.Uri;
import android.provider.ContactsContract;

import com.alexstyl.specialdates.events.peopleevents.CustomEventType;
import com.alexstyl.specialdates.events.peopleevents.EventType;
import com.alexstyl.specialdates.events.peopleevents.StandardEventType;

class CustomEventProvider {

    private static final Uri CONTENT_URI = ContactsContract.Data.CONTENT_URI;
    private static final String[] PROJECTION = {
            ContactsContract.CommonDataKinds.Event.LABEL
    };
    private static final String SELECTION =
            "( " + ContactsContract.Data._ID + " = ?" +
                    " AND " + ContactsContract.Data.MIMETYPE + " = ? " +
                    " AND " + ContactsContract.Data.IN_VISIBLE_GROUP + " = 1" +
                    ")";

    private static final String SORT_ORDER = ContactsContract.CommonDataKinds.Event._ID + " LIMIT 1";

    private final ContentResolver resolver;

    CustomEventProvider(ContentResolver resolver) {
        this.resolver = resolver;
    }

    EventType getEventWithId(long deviceId) {
        Cursor cursor = resolver.query(CONTENT_URI, PROJECTION, SELECTION, new String[]{
                String.valueOf(deviceId),
                ContactsContract.CommonDataKinds.Event.CONTENT_ITEM_TYPE
        }, SORT_ORDER);
        try {
            if (cursor.moveToFirst()) {
                int columnIndex = cursor.getColumnIndex(ContactsContract.CommonDataKinds.Event.LABEL);
                String eventName = cursor.getString(columnIndex);
                return new CustomEventType(eventName);
            }
        } finally {
            cursor.close();
        }
        return StandardEventType.OTHER;
    }
}

File: android_mobile/src/main/java/com/alexstyl/specialdates/events/peopleevents/CustomEventProvider.kt
package com.alexstyl.specialdates.events.peopleevents

import android.content.ContentResolver
import android.provider.ContactsContract

class CustomEventProvider(private val resolver: ContentResolver) {

    fun getEventWithId(deviceId: Long): EventType {

        resolver
                .query(CONTENT_URI,
                        PROJECTION,
                        SELECTION,
                        arrayOf(deviceId.toString(), ContactsContract.CommonDataKinds.Event.CONTENT_ITEM_TYPE),
                        SORT_ORDER)
                .use { cursor ->
                    if (cursor.moveToFirst()) {
                        val columnIndex = cursor.getColumnIndex(ContactsContract.CommonDataKinds.Event.LABEL)
                        val eventName = cursor.getString(columnIndex) ?: ""
                        return CustomEventType(eventName)
                    }
                }
        return StandardEventType.OTHER
    }

    companion object {

        private val CONTENT_URI = ContactsContract.Data.CONTENT_URI
        private val PROJECTION = arrayOf(ContactsContract.CommonDataKinds.Event.LABEL)
        private const val SELECTION = (
                "( " + ContactsContract.Data._ID + " = ?"
                        + " AND " + ContactsContract.Data.MIMETYPE + " = ? "
                        + " AND " + ContactsContract.Data.IN_VISIBLE_GROUP + " = 1"
                        + ")")

        private const val SORT_ORDER = ContactsContract.CommonDataKinds.Event._ID + " LIMIT 1"
    }
}
